File: lib/domain/notation/move_notation.dart
import '../models/board_point.dart';
import '../models/move.dart';

/// Phase 1 uses plain coordinate notation (file a-i, rank 0-9) rather than
/// traditional Chinese piece-relative notation, which needs disambiguation
/// rules (front/middle/back soldier, etc) out of scope for this phase.
class MoveNotation {
  MoveNotation._();

  static String squareLabel(BoardPoint p) => '${String.fromCharCode(97 + p.col)}${p.row}';

  static String forMove(Move move, {bool isCheck = false, bool isCheckmate = false}) {
    final separator = move.isCapture ? 'x' : '-';
    final base = '${squareLabel(move.from)}$separator${squareLabel(move.to)}';
    if (isCheckmate) return '$base#';
    if (isCheck) return '$base+';
    return base;
  }

  /// Parses a bare square label like "e3" (no separator) - the format UCI
  /// engines such as Pikafish use, which happens to match [squareLabel]'s
  /// own alphabet (file a-i, rank 0-9).
  static BoardPoint parseSquare(String label) {
    if (label.length < 2) {
      throw FormatException('Invalid square label "$label"');
    }
    final col = label.codeUnitAt(0) - 97; // 'a' = 0
    final row = int.parse(label.substring(1));
    return BoardPoint(row, col);
  }

  /// Parses a UCI coordinate move such as "e3e4" (always exactly two
  /// concatenated square labels - Xiangqi has no promotion suffix).
  static (BoardPoint from, BoardPoint to) parseUciMove(String move) {
    if (move.length != 4) {
      throw FormatException('Invalid UCI move "$move"');
    }
    return (parseSquare(move.substring(0, 2)), parseSquare(move.substring(2, 4)));
  }
}

File: backend/src/xiangqi/moveNotation.ts
import { BoardPoint, point } from './boardPoint';
import { isCapture, Move } from './move';

/** Ported from `lib/domain/notation/move_notation.dart`. Plain coordinate
 * notation (file a-i, rank 0-9) - this happens to be the same alphabet UCI
 * engines like Pikafish use for their coordinate moves, which is why
 * `parseSquare`/`parseUciMove` round-trip against it. */

export function squareLabel(p: BoardPoint): string {
  return `${String.fromCharCode(97 + p.col)}${p.row}`;
}

export function forMove(move: Move, opts: { isCheck?: boolean; isCheckmate?: boolean } = {}): string {
  const separator = isCapture(move) ? 'x' : '-';
  const base = `${squareLabel(move.from)}${separator}${squareLabel(move.to)}`;
  if (opts.isCheckmate) return `${base}#`;
  if (opts.isCheck) return `${base}+`;
  return base;
}

export function parseSquare(label: string): BoardPoint {
  if (label.length < 2) throw new Error(`Invalid square label "${label}"`);
  const col = label.charCodeAt(0) - 97; // 'a' = 0
  const row = Number.parseInt(label.slice(1), 10);
  return point(row, col);
}

export function parseUciMove(move: string): { from: BoardPoint; to: BoardPoint } {
  if (move.length !== 4) throw new Error(`Invalid UCI move "${move}"`);
  return { from: parseSquare(move.slice(0, 2)), to: parseSquare(move.slice(2, 4)) };
}
